Extend alarms API to support child devices

PradeepKiruvale opened the pull request:
## Proposed changes

- Extend the alarms feature to child/external devices.
- Create the child device, if it does not exists, and then forward the `alarm` message.
- Added tests to check the basic feature as well as to test the `reconciliation` of alarm messages.

## Types of changes

<!--
What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_
-->

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
https://github.com/thin-edge/thin-edge.io/issues/1204

## Checklist

<!--
_Put an `x` in the boxes that apply. You can also fill these out after
creating the PR. If you're unsure about any of them, don't hesitate to ask.
We're here to help! This is simply a reminder of what we are going to look for
before merging your code._
-->

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

matthiasbeyer commented on crates/core/tedge_mapper/src/c8y/converter.rs:
```diff
@@ -263,6 +267,38 @@ where
+                    ));
+                }
+            }
+            let () = self.size_threshold.validate(message)?;
```
```suggestion
            self.size_threshold.validate(message)?;
```

matthiasbeyer commented on crates/core/tedge_mapper/src/c8y/converter.rs:
```diff
@@ -574,13 +626,15 @@ impl AlarmConverter {
                     .for_each(|(_key, message)| sync_messages.push(message));
             }
             Self::Synced => {
+
```
```suggestion
```

PradeepKiruvale replied:
fixed

matthiasbeyer commented on crates/core/tedge_mapper/src/c8y/converter.rs:
```diff
@@ -574,13 +626,15 @@ impl AlarmConverter {
         sync_messages
     }
 }
+
```
```suggestion
```

PradeepKiruvale replied:
fixed

matthiasbeyer commented on crates/core/tedge_mapper/src/c8y/tests.rs:
```diff
@@ -372,6 +372,61 @@ async fn c8y_mapper_alarm_mapping_to_smartrest() {
+        .await
+        .unwrap();
+
+    let _ = broker
+        .publish_with_opts(
```
```suggestion
    broker.publish_with_opts(
```

PradeepKiruvale replied:
My editor indentation automatically changes to the previous way.

matthiasbeyer replied:
What editor and what plugins are you using?

PradeepKiruvale replied:
vscode rust assist for formating.

matthiasbeyer replied:
On rust toolchain for 1.58.1?

matthiasbeyer replied:
Okay, I just checked: rustfmt for 1.58.1 does not do this. And "rust assist" in vscode seems to use rustfmt under the hood. Any chance you're using some other plugin as well? 

Btw why are you using "rust assist" instead of [the official rust plugin for vscode](https://marketplace.visualstudio.com/items?itemName=rust-lang.rust)?

PradeepKiruvale replied:
Other plugins I have are "rust analyzer and rust syntax".

matthiasbeyer commented on crates/core/tedge_mapper/src/c8y/tests.rs:
```diff
@@ -372,6 +372,61 @@ async fn c8y_mapper_alarm_mapping_to_smartrest() {
+    .await;
+
+    //Clear the previously published alarm
+    let _ = broker
+        .publish_with_opts(
```
```suggestion
    broker.publish_with_opts(
```

PradeepKiruvale replied:
My editor indentation automatically changes to the previous way.

matthiasbeyer commented on crates/core/tedge_mapper/src/c8y/tests.rs:
```diff
@@ -462,6 +517,108 @@ async fn c8y_mapper_syncs_pending_alarms_on_startup() {
+        .messages_published_on("c8y-internal/alarms/critical/temperature_alarm/external_sensor")
+        .await;
+
+    let _ = broker
+        .publish_with_opts(
```
```suggestion
    broker.publish_with_opts(
```

matthiasbeyer commented on crates/core/tedge_mapper/src/c8y/tests.rs:
```diff
@@ -462,6 +517,108 @@ async fn c8y_mapper_syncs_pending_alarms_on_startup() {
+    sm_mapper.abort();
+
+    //Publish a new alarm while the mapper is down
+    let _ = broker
+        .publish_with_opts(
```
```suggestion
    broker.publish_with_opts(
```

matthiasbeyer commented on crates/core/tedge_mapper/src/c8y/tests.rs:
```diff
@@ -462,6 +517,108 @@ async fn c8y_mapper_syncs_pending_alarms_on_startup() {
+        .await
+        .unwrap();
+
+    let _ = broker
+        .publish_with_opts(
```
```suggestion
    broker.publish_with_opts(
```

matthiasbeyer commented on crates/core/tedge_mapper/src/c8y/converter.rs:
```diff
@@ -313,13 +349,11 @@ where
+                if topic.name.starts_with("tedge/alarms")
+                    | topic.name.starts_with(INTERNAL_ALARMS_TOPIC) =>
+            {
+                Ok(self.process_alarm_messages(topic, message)?)
```
```suggestion
                self.process_alarm_messages(topic, message)
```

It seems that this function does return the same type as `process_alarm_messages()`, so I guess this should work.

PradeepKiruvale replied:
Yes, it works.

matthiasbeyer commented on crates/core/tedge_mapper/src/c8y/converter.rs:
```diff
@@ -479,11 +515,16 @@ impl AlarmConverter {
             }
             Self::Synced => {
                 //Regular conversion phase
+
```
```suggestion
```

PradeepKiruvale replied:
removed

matthiasbeyer commented on crates/core/tedge_mapper/src/c8y/converter.rs:
```diff
@@ -479,11 +515,16 @@ impl AlarmConverter {
+
                 let tedge_alarm =
                     ThinEdgeAlarm::try_from(input.topic.name.as_str(), input.payload_str()?)?;
+
```
```suggestion
```

matthiasbeyer commented on crates/core/tedge_mapper/src/c8y/tests.rs:
```diff
@@ -462,6 +517,108 @@ async fn c8y_mapper_syncs_pending_alarms_on_startup() {
+
+    // Ignored until the rumqttd broker bug that doesn't handle empty retained messages
+    //Clear the existing alarm while the mapper is down
+    // let _ = broker
+    //     .publish_with_opts(
```
```suggestion
    // broker.publish_with_opts(
```

matthiasbeyer reviewed (changes requested):
I had a quick look and found some style things.

Also, these commits should be squashed:

* f484f8274d6ef26d42779ccb9e034a6236d0025e
* 9cb4998446e4963f204fde27fdd34b3e4322a7e6
* ba729e478db55ce59cb1cc6445cea919db15f771

And the commit messages of these should be rewritten to something more explanatory:

* e1c5d0b192942b922f826874583749ab0b457adf
* b16bd094ccf31ee784acdb98a67c990183b19657

:wink:

rina23q commented on crates/core/tedge_mapper/src/c8y/converter.rs:
```diff
@@ -461,9 +495,11 @@ impl AlarmConverter {
-
+    fn try_convert_alarm(
+        &mut self,
+        mqtt_messages: &mut Vec<Message>,
```
I really got confused with the two arguments: `mqtt_messages` and `input` since both are `Message`. I think this `mqtt_messages` should be renamed to `output_messages` because MQTT is very obvious from the type and good contrast with `input`.

```suggestion
        output_messages: &mut Vec<Message>,
```

albinsuresh replied:
Instead of passing the vector of `output_messages` into this function, I'd rather append the result of the this function to the messages vector in the calling function: `process_alarm_messages`.

rina23q reviewed (changes requested):
Many problems on handling corner cases.

**Case 1. Publish empty payload to an invalid topic for parent device `tedge/alarms/major/`**
- Before(Expected): It returned an error on `tedge/errors`
```
[tedge/alarms/major/] 
[tedge/errors] Unsupported topic: tedge/alarms/major/
```
- Actual: It doesn't return any error.
```
[tedge/alarms/major/] 
```

**Case 2. Publish empty payload to an invalid topic for child device `tedge/alarms/major//childfoo`**
- Expected: Error is published on `tedge/errors`.
- Actual: It doesn't return any error. 

**Case 3. Publish empty payload to an invalid topic for child device `tedge/alarms/minor/aaa/`**
- Expected: Error is published `tedge/errors` as empty child name is not acceptable.
- Actual: It accepts, then Cumulocity returns an error.
```
[tedge/alarms/minor/aaa/] 
[c8y/s/us] 101,,,thin-edge.io-child
[c8y/s/us/] 306,aaa
[c8y-internal/alarms/minor/aaa/] 
[c8y/s/e] 41,101,requires '1' value: childSerial
```

I tested with an empty payload to simplify the tests, however, all problems are occurred by the insufficient validation of `tedge/alarms/...`. So, I assume, even with payload, the same problems will happen. 
I believe, if you fix the corresponding function, everything will be solved.

Also another problem is lacking of such unit tests, to check lots of patterns of topic name.

PradeepKiruvale commented:
> I had a quick look and found some style things. Some of those are simply changes that are unrelated.
> 
> Also, these commits should be squashed into the others:
> 
> * [f484f82](https://github.com/thin-edge/thin-edge.io/commit/f484f8274d6ef26d42779ccb9e034a6236d0025e)
> * [9cb4998](https://github.com/thin-edge/thin-edge.io/commit/9cb4998446e4963f204fde27fdd34b3e4322a7e6)
> * [ba729e4](https://github.com/thin-edge/thin-edge.io/commit/ba729e478db55ce59cb1cc6445cea919db15f771)
> 
> And the commit messages of these should be rewritten to something more explanatory:
> 
> * [e1c5d0b](https://github.com/thin-edge/thin-edge.io/commit/e1c5d0b192942b922f826874583749ab0b457adf)
> * [b16bd09](https://github.com/thin-edge/thin-edge.io/commit/b16bd094ccf31ee784acdb98a67c990183b19657)
> 
> 😉

I will squash them before merging. I just kept them separate to make it easy for the reviewer.

albinsuresh commented on crates/core/thin_edge_json/src/alarm.rs:
```diff
@@ -63,7 +63,7 @@ impl ThinEdgeAlarm {
     ) -> Result<Self, ThinEdgeJsonDeserializerError> {
         let topic_split: Vec<&str> = mqtt_topic.split('/').collect();
-        if topic_split.len() == 4 {
+        if topic_split.len() == 4 || topic_split.len() == 5 {
```
> If user publishes onto tedge/alarms/major/temperature_alarm/ (after / is empty), is it an error? Or we should treat it as an alarm for parent device?

I'd say it's an error.

PradeepKiruvale replied:
> You should add unit tests for
> 
> * More than 5 (e.g. topic is `tedge/alarms/major/temperature_alarm/foo/bar`)
> * Just 5 with valid child name (e.g. topic is `tedge/alarms/major/temperature_alarm/foo`)
> * Just 5 with invalid child name (e.g. topic is `tedge/alarms/major/temperature_alarm/`)
> 
> But actually the last case is a question. If user publishes onto `tedge/alarms/major/temperature_alarm/` (after `/` is empty), is it an error? Or we should treat it as an alarm for parent device?

Added unit tests

albinsuresh commented on crates/core/tedge_mapper/src/c8y/converter.rs:
```diff
@@ -263,6 +267,38 @@ where
+            let mut mqtt_messages: Vec<Message> = Vec::new();
+            let topic_split: Vec<&str> = topic.name.split('/').collect();
+            if topic_split.len() == 5 {
+                let c_id = topic_split[4];
```
Missed this one?

PradeepKiruvale commented on crates/core/tedge_mapper/src/c8y/tests.rs:
```diff
@@ -372,6 +372,61 @@ async fn c8y_mapper_alarm_mapping_to_smartrest() {
+    .await;
+
+    //Clear the previously published alarm
+    let _ = broker
```
That is to clear the alarm message, validation is not needed here.

rina23q reviewed (commented):
All bugs were fixed :+1: 
Need little work on code.

PradeepKiruvale commented on crates/core/thin_edge_json/src/alarm.rs:
```diff
@@ -39,6 +39,9 @@ pub enum ThinEdgeJsonDeserializerError {
     #[error(transparent)]
     SerdeJsonError(#[from] serde_json::error::Error),
+
+    #[error("Unsupported external device in topic: {0}")]
```
Yeah, that's right.

PradeepKiruvale commented on crates/core/thin_edge_json/src/alarm.rs:
```diff
@@ -215,4 +247,46 @@ mod tests {
     }
+
+    #[test]
+    fn alarm_translation_empty_child_name() {
```
fixed

PradeepKiruvale commented on crates/core/thin_edge_json/src/alarm.rs:
```diff
@@ -215,4 +247,46 @@ mod tests {
+    #[test]
+    fn alarm_translation_empty_child_name() {
+        let result = ThinEdgeAlarm::try_from("tedge/alarms/critical/temperature_alarm//", "{}");
+        println!("{:?}", result);
```
Removed

PradeepKiruvale commented:
> I had a quick look and found some style things. Some of those are simply changes that are unrelated.
> 
> Also, these commits should be squashed into the others:
> 
> * [f484f82](https://github.com/thin-edge/thin-edge.io/commit/f484f8274d6ef26d42779ccb9e034a6236d0025e)
> * [9cb4998](https://github.com/thin-edge/thin-edge.io/commit/9cb4998446e4963f204fde27fdd34b3e4322a7e6)
> * [ba729e4](https://github.com/thin-edge/thin-edge.io/commit/ba729e478db55ce59cb1cc6445cea919db15f771)
> 
> And the commit messages of these should be rewritten to something more explanatory:
> 
> * [e1c5d0b](https://github.com/thin-edge/thin-edge.io/commit/e1c5d0b192942b922f826874583749ab0b457adf)
> * [b16bd09](https://github.com/thin-edge/thin-edge.io/commit/b16bd094ccf31ee784acdb98a67c990183b19657)
> 
> 😉
Now I have squashed my commits to one commit and added more explanation about this feature.

PradeepKiruvale commented:
#1204
